### ISSUE TITLE
Rerun of flaky tests failed

### ISSUE THREAD

**Issue by henning-krause** [Microsoft/vsts-tasks#7037]
## Environment
- Server - TFS 2018.2rc on-premises
- Agent - Private, 2.131.0 on Windows Server 2016

## Issue Description

Task         : Visual Studio Test
Version      : 2.3.27

I've enabled to rerun failed tests. During a particular test run, I got this error during test execution:

```
2018-04-20T13:29:51.6833758Z ##[error]Incorrect format for TestCaseFilter Missing Operator '|' or '&'. Specify the correct format and try again. Note that the incorrect format can lead to no test getting executed.
```

This error occurred for each test assembly that was specified in the vstest command (at least the number this error was repeated is the same as the number of assemblies being tested.

This was the command that was executed:

```
vstest.console.exe /TestCaseFilter:"FullyQualifiedName=SomeNamespace.Tests.Test1  (False)"
```

Is it possible, that the TestCaseFilter does not work parameterized unit tests?

**Comment by nigurr** [comment]
@henning-krause 
Hi,

Currently we don't support data driven tests for Rerun scenario. It will be available with TFS 2019 release.

**Comment by Galad** [comment]
@nigurr 
Hello,
I observered the same issue on VSTS.
Does your answer counts for VSTS as well ?

**Comment by nigurr** [comment]
@Galad Yes, this fix requires VS 15.8 alternatively you can add "Visual Studio Platform Installer" tool task before VsTest task
![image](https://user-images.githubusercontent.com/15208679/40104472-0e294d70-590e-11e8-8556-6c643561d657.png)

**Comment by Galad** [comment]
Thanks

**Comment by tebeco** [comment]
Vs15.8 will also fix this for `Live testing` ?
I cannot trust my coverage anymore ^^

**Comment by mattpickle** [comment]
@nigurr We are using `VsTest Platform Installer` which installs 15.8.0, but still getting errors on reruns for our SpecFlow/Nunit tests.  Any suggestions?

```
2018-08-02T15:45:48.0238928Z **************** Rerunning failed tests for Test run 54 *********************
2018-08-02T15:45:48.0239487Z Max attempts: 2; Current attempt: 1; Failed test cases threshold: 30; Failed test cases max limit: 0
2018-08-02T15:45:48.0240044Z [command]C:\Users\test\Downloads\vsts-agent-win-x64-2.136.1\_work\_tool\VsTest\15.8.0\x64\tools\net451\Common7\IDE\Extensions\TestPlatform\vstest.console.exe @C:\Users\test\AppData\Local\Temp\tmp3577.tmp
2018-08-02T15:45:48.1565474Z Microsoft (R) Test Execution Command Line Tool Version 15.8.0
2018-08-02T15:45:48.1567177Z Copyright (c) Microsoft Corporation.  All rights reserved.
2018-08-02T15:45:48.1654700Z vstest.console.exe /TestCaseFilter:"FullyQualifiedName=Company.Automation.Features.TestFeatures.HTML5Viewer.Annotations.AnnotationsDragDropFeature.VerifyAnnotations_DragDrop(\"Automation-NonRCRT\"\,\"Documents\\\\1.tiff\"\,\"Ellipse\"\,\"Documents\\\\1.VerifyAnnotations_DragDrop(\"Automation-NonRCRT\"\,\"Documents\\\\1.tiff\"\,\"Ellipse\"\,\"Documents\\\\1.tiff\"\,[\"annotations-ellipse\"])|FullyQualifiedName=Company.Automation.Features.TestFeatures.HTML5Viewer.Annotations.AnnotationsDragDropFeature.VerifyAnnotations_DragDrop(\"Automation-NonRCRT\"\,\"Documents\\\\1.tiff\"\,\"Ellipse\"\,\"Documents\\\\1.VerifyAnnotations_DragDrop(\"Automation-NonRCRT\"\,\"Documents\\\\1.tiff\"\,\"Ellipse\"\,\"Documents\\\\1.tiff\"\,[\"annotations-ellipse\"])"
2018-08-02T15:45:48.1657017Z "C:\Users\test\Downloads\vsts-agent-win-x64-2.136.1\_work\1\s\Company.Automation.Features\bin\Debug\Company.Automation.Features.dll"
2018-08-02T15:45:48.1658083Z "C:\Users\test\Downloads\vsts-agent-win-x64-2.136.1\_work\1\s\Company.Automation.Features\obj\Debug\Company.Automation.Features.dll"
2018-08-02T15:45:48.1658834Z /Settings:"C:\Users\test\Downloads\vsts-agent-win-x64-2.136.1\_work\_temp\e4ad1140-9669-11e8-829e-c752c11429f2.runsettings"
2018-08-02T15:45:48.1659437Z /InIsolation
2018-08-02T15:45:48.1659888Z /logger:"trx"
2018-08-02T15:45:48.1660472Z /TestAdapterPath:"C:\Users\test\Downloads\vsts-agent-win-x64-2.136.1\_work\1\s"
2018-08-02T15:45:48.6087462Z Starting test execution, please wait...
2018-08-02T15:45:49.6242718Z Multiple versions of same extension found. Selecting the highest version.
2018-08-02T15:45:49.6243646Z   NUnit3.TestAdapter : 3.10.0.21
2018-08-02T15:45:49.6243915Z Multiple versions of same extension found. Selecting the highest version.
2018-08-02T15:45:49.6244066Z   NUnit3.TestAdapter : 3.10.0.21
2018-08-02T15:45:50.0697913Z NUnit Adapter 3.10.0.21: Test execution started
2018-08-02T15:45:50.0698225Z NUnit Adapter 3.10.0.21: Test execution started
2018-08-02T15:45:50.0712799Z ##[error]An exception occurred while invoking executor 'executor://nunit3testexecutor/': Filter string '"Automation-NonRCRT"\,"Documents\\\\1.tiff"\,"Ellipse"\,"Documents\\\\1.VerifyAnnotations_DragDrop' includes unrecognized escape sequence.
2018-08-02T15:45:50.0713824Z ##[error]An exception occurred while invoking executor 'executor://nunit3testexecutor/': Filter string '"Automation-NonRCRT"\,"Documents\\\\1.tiff"\,"Ellipse"\,"Documents\\\\1.VerifyAnnotations_DragDrop' includes unrecognized escape sequence.
2018-08-02T15:45:50.2680384Z ##[error]Test Run Failed.
```

**Comment by yanrez** [comment]
@nigurr can you please explain why setting batch size helps and how it related to the option to rerun failed tests?

**Comment by acesiddhu** [comment]
When we set a batch size in the task, task runs the test by calling into API of the test platform instead of calling vstest.console.exe command line. The problem with command line is it doesn't work well with test cases with , . spaces etc.

**Comment by Thilas** [comment]
@nigurr, @acesiddhu, I also have this issue (using VS 15.9 release) and it seems to me that the cause is some bad escaping of parentheses (at least in my cases). If I run manually the following command line: `vstest.console.exe TestLibrary.dll /TestCaseFilter:"FullyQualifiedName=TestLibrary.TestClass.Test(1)"`, then it failed. But if I run: `vstest.console.exe TestLibrary.dll /TestCaseFilter:"FullyQualifiedName=TestLibrary.TestClass.Test\(1\)"`, it succeeds.

In the meantime, I would like to use to provided workaround (setting batch size) but I have a question first: is it recommended even when there are like thousands of tests (or even more)?

Thanks.

**Comment by nigurr** [comment]
@Thilas There shouldn't be an issue with using batching option. If you are using single agent to run all tests, we recommend to set the value to total number of tests you have, in such case all tests will be executed in one shot without much delay. Functionality wise, both options offers the same.

**Comment by Thilas** [comment]
@nigurr, ok I'll try then. Thanks.

**Comment by mirashgoku** [comment]
I am not sure if this its the right place to ask, but here goes nothing

I have been running automated testing using selenium Webdrivers on VSTS Build on a private agent every night and for the last 3 weeks i have been facing this issue:

![image](https://user-images.githubusercontent.com/22146975/52195076-9d211400-2856-11e9-83fa-ea03b07c512c.png)

I have tried everything that came to my mind, and also looked online for it with no success. 

I have tried the steps above, but ithe test got aborted:

![image](https://user-images.githubusercontent.com/22146975/52195302-a1016600-2857-11e9-9a70-0a411ec68e12.png)

I am not sure what i should check next. I ask for guidance.

Thanks
